platform/osgi/platform-osgi-condpermadmin/src/main/java/org/codice/ddf/condpermadmin/PermissionActivator.java

```diff
@@ -79,47 +90,114 @@ public void start(BundleContext bundleContext) throws Exception {
       conditionalPermissionUpdate.getConditionalPermissionInfos().addAll(allDenyInfos);
       conditionalPermissionUpdate
           .getConditionalPermissionInfos()
-          .add(getAllPermission(conditionalPermissionAdmin, ConditionalPermissionInfo.ALLOW));
+          .add(getAllPermission(ConditionalPermissionInfo.ALLOW));
     } else if (priorityResult == Priority.DENY) {
       conditionalPermissionUpdate.getConditionalPermissionInfos().addAll(allDenyInfos);
       conditionalPermissionUpdate.getConditionalPermissionInfos().addAll(allGrantInfos);
       conditionalPermissionUpdate
           .getConditionalPermissionInfos()
-          .add(getAllPermission(conditionalPermissionAdmin, ConditionalPermissionInfo.DENY));
+          .add(getAllPermission(ConditionalPermissionInfo.DENY));
     }
 
     conditionalPermissionUpdate.commit();
+    final Dictionary<String, Object> props = new Hashtable<>(8);
+
+    props.put(Constants.SERVICE_DESCRIPTION, "DDF :: Platform :: OSGi :: CondPermAdmin");
+    props.put(Constants.SERVICE_VENDOR, "Codice Foundation");
+    bundleContext.registerService(PermissionService.class, this, props);
+  }
+
+  @Override
+  public void grantPermission(String bundle, String permission) throws Exception {
+    synchronized (this) {
+      // use the parsed policy to make it easier to parse the permission string
+      final ParsedPolicy parsedPolicy =
+          new Parser(false)
+              .parse(
+                  new StringReader(
+                      String.format(
+                          "grant codebase \"file:/%s\" { permission %s; }", bundle, permission)));
+      final List<ParsedPolicyEntry> grantEntries = parsedPolicy.getGrantEntries();
+      final List<ConditionalPermissionInfo> allGrantInfos = new ArrayList<>();
+      final ConditionalPermissionUpdate conditionalPermissionUpdate =
+          conditionalPermissionAdmin.newConditionalPermissionUpdate();
+
+      buildConditionalPermissionInfo(grantEntries, allGrantInfos, ConditionalPermissionInfo.ALLOW);
+      final ConditionalPermissionInfo grantInfo = allGrantInfos.get(0);
+      final List<ConditionalPermissionInfo> conditionalInfos =
+          conditionalPermissionUpdate.getConditionalPermissionInfos();
+      boolean added = false;
+
+      // see if we can find one conditional permission for the exact same permission
+      // if we do, let's just add this new bundle to the list as opposed to adding a
+      // brand new conditional permission
+      for (final ListIterator<ConditionalPermissionInfo> i = conditionalInfos.listIterator();
+          i.hasNext(); ) {
+        final ConditionalPermissionInfo permInfo = i.next();
+
+        if (Objects.equals(grantInfo.getAccessDecision(), permInfo.getAccessDecision())
+            && Arrays.equals(grantInfo.getPermissionInfos(), permInfo.getPermissionInfos())) {
+          final ConditionInfo[] conditions = permInfo.getConditionInfos();
+
+          if ((conditions != null)
```

Comment:
✏️ I think this loop would be a bit faster if you switched the order of the conditional tests. Get the `conditions` at the top of the loop and check that it's not null, length 1, and a BundleNameCondition before performing the equality tests.

Reply:
I cannot really do that since `conditions` come from the loop index dependent `permInfo`.

Reply:
Yeah. Move line 169 to line 166 and then execute the null/length/bundleName test before the equality tests. That should work unless I'm just blind this morning.

Reply:
I just didn't want to look at the conditions until I knew it was the right access and permission info to start with. The fact is that there are a lot less changes of that conditions array being null or not of size 1 once the above 2 are met. The tests are include anyway for completeness and protection.

Reply:
I get that, but assuming `getConditionInfos()` is a simple getter and not performing any logic to generate the array, checking its meta information is going to be cheaper than those equality tests. It's not likely a huge performance difference, but is non-zero.

Unless my assumption is wrong and `getConditionInfos()` is an expensive operation; in that case, this suggestion is definitely wrong.

Reply:
The same comment could be said about `getConditionInfos()` which is why I structured it this way. But I agree with you, this one was not a clear cut one where typically I will also try to do the simpler tests first.